sdks/python/apache_beam/ml/transforms/handlers.py

```diff
@@ -0,0 +1,431 @@
+      row_type = RowTypeConstraint.from_user_type(element_type)
+      if not row_type:
+        raise TypeError(
+            "Element type must be compatible with Beam Schemas ("
+            "https://beam.apache.org/documentation/programming-guide/#schemas)"
+            " for to use with MLTransform and TFTProcessHandlerSchema.")
+    else:
+      row_type = element_type
+    inferred_types = {name: typ for name, typ in row_type._fields}
+
+    for k, t in inferred_types.items():
+      if t in _primitive_types_to_typing_container_type:
+        inferred_types[k] = _primitive_types_to_typing_container_type[t]
+
+    # sometimes a numpy type can be provided as np.dtype('int64').
+    # convert numpy.dtype to numpy type since both are same.
+    for name, typ in inferred_types.items():
+      if isinstance(typ, np.dtype):
+        inferred_types[name] = typ.type
+
+    return inferred_types
+
+  def process_data_fn(
+      self, inputs: Dict[str, common_types.ConsistentTensorType]
+  ) -> Dict[str, common_types.ConsistentTensorType]:
+    """
+    This method is used in the AnalyzeAndTransformDataset step. It applies
+    the transforms to the `inputs` in sequential order on the columns
+    provided for a given transform.
+    Args:
+      inputs: A dictionary of column names and data.
+    Returns:
+      A dictionary of column names and transformed data.
+    """
+    outputs = inputs.copy()
+    for transform in self.transforms:
+      columns = transform.columns
+      for col in columns:
+        intermediate_result = transform.apply(
+            outputs[col], output_column_name=col)
+        for key, value in intermediate_result.items():
+          outputs[key] = value
+    return outputs
+
+  def _get_transformed_data_schema(
+      self,
+      metadata: dataset_metadata.DatasetMetadata,
+  ) -> Dict[str, typing.Sequence[typing.Union[np.float32, np.int64, bytes]]]:
+    schema = metadata._schema
+    transformed_types = {}
+    logging.info("Schema: %s", schema)
+    for feature in schema.feature:
+      name = feature.name
+      feature_type = feature.type
+      if feature_type == schema_pb2.FeatureType.FLOAT:
+        transformed_types[name] = typing.Sequence[np.float32]
+      elif feature_type == schema_pb2.FeatureType.INT:
+        transformed_types[name] = typing.Sequence[np.int64]
+      elif feature_type == schema_pb2.FeatureType.BYTES:
+        transformed_types[name] = typing.Sequence[bytes]
+      else:
+        # TODO: This else condition won't be hit since TFT doesn't output
+        # other than float, int and bytes. Refactor the code here.
+        raise RuntimeError(
+            'Unsupported feature type: %s encountered' % feature_type)
+    logging.info(transformed_types)
+    return transformed_types
+
+  def process_data(
+      self, pcoll: beam.PCollection[tft_process_handler_schema_input_type]
+  ) -> beam.PCollection[beam.Row]:
+
+    self._fail_on_non_gloabl_window(pcoll)
```

Comment:
This should only be the case for artifact production. We want to support artifact consumption operations in streaming mode.

Reply:
Some transforms provide artifacts such as `tft.scale_to_0_1`. Some transforms consume https://www.tensorflow.org/tfx/transform/api_docs/python/tft/scale_by_min_max which takes the artifacts produced. 

So should I just check the transforms the user provided and determine if we need to fail the pipeline or not?

Reply:
I thought all transforms had two modes, a transform-producing and transform-consuming variant. E.g. tft.scale_by_min_max in "producing" mode does a pass over the data to figure out what the min and max of the column is (to properly scale the data), and in consuming mode takes the previously-computed min and max to scale the data.

Reply:
tft has two sets of transforms: [analyzers](https://github.com/tensorflow/transform/blob/master/tensorflow_transform/analyzers.py) and [mappers](https://github.com/tensorflow/transform/blob/master/tensorflow_transform/mappers.py). Analyzers typically do reductions (i.e. produce "artifact"), mappers can be either pure (don't need a reduction) or requiring analyzers. Analyzers basically produce a "placeholder" that is placed into a produced tf graph and is refereed to in the mappers. So the first thing we do is run the analysis stage that substitutes these placeholders with the reduced value as a constant. The artifact that tft produces is basically the tf graph (with assets) with these placeholders substituted with constants produced during analysis stage.

The downstream streaming variant just applies these transformations as if the reduced values were constants in the produced graph.

Reply:
tft.scale_by_min_max if used in the `preprocessing_fn` that is passed to `AnalyzeandTransformDataset` (or `AnalyzeDataset`) will produce the min and max values in the artifact. [Output artifact](https://github.com/tensorflow/transform/blob/e6e2a6890572287d7900a37c8c40e03826c78b41/tensorflow_transform/output_wrapper.py#L60) of such run when appliead to data will do the scaling with the previously produced values.

We have no out-of-the-box way to use the produced min and max in any different way (i.e. to do something other than scaling)

Reply:
As a non tensorflow user, if I use `tft.scale_by_min_max` on dataset A and got `min` and `max` value of the dataset A after preprocessing the entire data.  Later if I want to use the same `tft.scale_by_min_max` on dataset B using the `min`, 'max' values I got from dataset A, there is no way we can plug the artifacts we produced  previously right?

Reply:
@robertwb So we could ask the user about the artifact location. For preprocessing, we generate artifacts and store them in `artifact_location`. This is the `producing` stage.

For the consuming stage, we will ask the user to provide the same location, we will load the artifacts(transform_fn) and do the processing with produced artifacts instead of computing them again.

Reply:
Correct. How do we distinguish between the two modes?

Reply:
We ask for the `artifact_location`. Validate `artifact_location` using `tft.TFTransformOutput`. If it fails, we assume we should compute artifacts. If the validation is successful, we load the `artifacts` and plug them in to the pipeline.

We can also ask the users explicitly with input parameters to TFTProcessHandler.

Reply:
This seems rather risky to infer. Consider the cases

(1) In iterating on my algorithm, I run over larger and larger portions of my actual data. All the constants are fixed on my first run and never learn about the new data.

(2) Due to corruption or some other mismatch, the desired artifact fails to load. I silently compute new artifacts and try to apply these different normalizations to a model trained with the prior normalizations.  

I think we need to be explicit here.

Reply:
In the consuming mode users don't call `tft.scale_by_min_max` -- it is already embedded in the artifact with the scaling values. So the just load the whole artifact and apply it to the same (or a subset) of features and get the transformations in the consuming mode.

If you need actually need the min and max values (and not just the callable scaler):
Although TFT doesn't always store the reduction values separately (it stores quantiles and vocab separately), there's a pretty clear path of how that could be done, so if you think this is necessary - we would support a contribution.

If you want to avoid changes in TFT, there's a hacky workaround as well: e.g. for `tft.scale_by_min_max` you could apply the model to [-1, 1] getting [res0, res1] and solve the system of equations:
(-1 - min)/(max - min) = res0
(1 - min)/(max-min) = res1